### DISCUSSION TITLE
FromStr vs. TryFrom

### DISCUSSION THREAD

**Discussion by ThomasdenH** [rust-lang-nursery/api-guidelines#182] [API Guidelines]
There are several issues related to this in the rust repo ([#1822](https://github.com/rust-lang/rfcs/issues/1822), [#2143](https://github.com/rust-lang/rfcs/issues/2143), [#53167](https://github.com/rust-lang/rust/issues/53167)). The FromStr and TryFrom traits are very similar. Which should a library implement?

**Comment by BurntSushi** [comment]
`TryFrom` [isn't stable yet](https://github.com/rust-lang/rust/issues/33417), so I think the answer for now is pretty easy: implement `FromStr`. :-)

**Comment by erickt** [comment]
TryFrom is queued up for [stabilization in 1.34](https://github.com/rust-lang/rust/pull/58302), so it might be time to start thinking about when to implement the two traits. In my opinion, I think it'd be okay to have both, and possibly implement one in terms of the other.

**Comment by ThomasdenH** [comment]
I think it would be best to implement both. API wise it doesn't matter which is implemented in terms of the other so no recommendation is necessary there.

**Comment by JeanMertz** [comment]
One reason to favour `TryFrom` (which by now has been stable for some time) is that `FromStr` cannot be implemented for a type with lifetime parameters.

**Comment by Kixunil** [comment]
I'd suggest that if a type implements `FromStr` it should also implement `TryFrom<String>` with the trivial implementation being `value.parse()`. The idea is to allow reusing the allocated string if the caller doesn't need to keep it. This is already possible for `String` (thanks to `From` implementations).

In other words, any type that implements one of them should implement all of them. Might be done using a macro.
